### PR TITLE
Video Sections

### LINKED ISSUE #339: Video sections
**Is your feature request related to a problem? Please describe.**
Course Admins should be allowed to add timestamps with section names for each lecture.

**Describe the solution you'd like**
Simple UI on the admin dashboard to add timestamps and section names. The sections will then be displayed beneath the VOD. A click on a section will jump to the timestamp in the VOD. For example, YouTube's solution looks like the following: 

image

### PR DESCRIPTION
**Pull request by MatthiasReumann**
- Closes #339 
- Video Sections can be created on the admin page for VOD lectures 
- Added navigation on Watch page to jump to given sections (If video is already running)

### CODE REVIEW

File: dao/video-section.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+func CreateVideoSectionBatch(sections []model.VideoSection) error {
+	var err error
+	for _, section := range sections {
+		err = CreateVideoSection(&section)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
Comment (joschahenningsen) [inline dao/video-section.go]: Why not do it like this?

```suggestion
func CreateVideoSections(sections []model.VideoSection) error {
	return DB.Create(&sections).Error
}
```

File: model/video-section.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+type VideoSection struct {
+	gorm.Model
+
+	Description    string `gorm:"not null" json:"description"`
+	StartHours uint   `gorm:"not null" json:"startHours"`
+	StartMinutes uint   `gorm:"not null" json:"startMinutes"`
+	StartSeconds uint   `gorm:"not null" json:"startSeconds"`
+
+	StreamID       uint   `gorm:"not null" json:"streamID"`
```
Comment (joschahenningsen) [inline model/video-section.go]: Can you `go fmt` this file?

File: web/template/watch.gohtml [hunk 3]
```diff
@@ -211,6 +215,41 @@
                         @click="less = !less" x-text="less ? 'show more' : 'show less'">show less
                 </button>
             </div>
+            <!-- Video Sections -->
+            {{ if $stream.IsPast }}
+                <div x-data="{sections: new watch.VideoSectionList({{$stream.ID}})}"
+                     x-init="await sections.fetch()"
```
Comment (joschahenningsen) [inline web/template/watch.gohtml]: Can the sections be provided in the DOM without requiring an extra api call? Do we really need them in alpine here or could the same be achieved using a plain `range` and sprinkle some alpine on top for the pages?

File: web/template/watch.gohtml [hunk 4]
```diff
@@ -211,6 +215,41 @@
                         @click="less = !less" x-text="less ? 'show more' : 'show less'">show less
                 </button>
             </div>
+            <!-- Video Sections -->
+            {{ if $stream.IsPast }}
```
Comment (joschahenningsen) [inline web/template/watch.gohtml]: Should this be `$stream.Recording`?

File: web/template/watch.gohtml [hunk 5]
```diff
@@ -211,6 +215,41 @@
                         @click="less = !less" x-text="less ? 'show more' : 'show less'">show less
                 </button>
             </div>
+            <!-- Video Sections -->
```
Comment (joschahenningsen) [inline web/template/watch.gohtml]: The section picker should probably get it's own file and be embedded here to improve readability

File: web/template/admin/admin_tabs/edit-course.gohtml [hunk 6]
```diff
@@ -200,56 +200,183 @@
...
+                                                            {{if $lecture.Stats}}
+                                                                <a>
+                                                                    <i class="fas fa-chart-line text-4 text-lg hover:text-1"
+                                                                       title="Live Statistics"
+                                                                       onclick="admin.showStats({{$lecture.Model.ID}})"></i>
+                                                                </a>
+                                                            {{end}}
+                                                        </div>
+
+                                                        <!-- create video sections -->
```
Comment (joschahenningsen) [inline web/template/admin/admin_tabs/edit-course.gohtml:231]: This might also get it's own template file

File: web/template/watch.gohtml (hunk 3 above)
Comment (joschahenningsen) [reply web/template/watch.gohtml]: E.g. rendering all of them, use `overflow-x-scroll` and create buttons to scroll the container

File: web/template/watch.gohtml [hunk 7]
```diff
@@ -211,6 +215,41 @@
...
+                                <i class="fa fa-chevron-left rotate-90 lg:rotate-0"></i>
+                            </button>
+                        </template>
+                        <template x-for="section in sections.currentGroupSections()">
+                            <button class="w-full lg:w-auto flex bg-transparent outline-none border-0"
+                                    @click="watch.jumpTo(section.startHours, section.startMinutes, section.startSeconds)">
+                                <div class="flex px-2 py-1 bg-white dark:bg-secondary justify-between w-full border dark:border-gray-800 hover:dark:bg-gray-600 hover:bg-gray-200 rounded-lg">
+                                    <span class="text-sm text-3 my-auto mr-3" x-text="section.description"></span>
+                                    <span class="text-sky-800 my-auto bg-sky-200 text-xs dark:text-indigo-200 dark:bg-indigo-800 p-1 rounded"
+                                          x-text="`${section.startHours}:${section.startMinutes}:${section.startSeconds}`"></span>
```
Comment (joschahenningsen) [inline web/template/watch.gohtml]: The timestamps should be left-padded with a 0 if the number is < 10

**Review by joschahenningsen — CHANGES_REQUESTED** [review]
This is amazing and works great! I just have a few little questions and remarks. 
Also, can the unrelated (formatting) changes here be removed to avoid merge conflicts with 
#389?

File: dao/video-section.go (hunk 1 above)
Comment (MatthiasReumann) [reply dao/video-section.go]: That's a pretty good question. Guess my brain wasn't working properly again. 😅 Thanks for pointing out!

File: web/template/watch.gohtml (hunk 4 above)
Comment (MatthiasReumann) [reply web/template/watch.gohtml]: Yup, thanks.

**Comment by MatthiasReumann** [comment]
In addition to the changes requested, I implemented a VideoSection interface as suggested in #388.

File: web/template/watch.gohtml (hunk 3 above)
Comment (MatthiasReumann) [reply web/template/watch.gohtml]: Yup, good idea. I moved the section to the right of the description with y-scroll. Looks even better imo and not a single line of alpine needed for pagination.

**Comment by joschahenningsen** [comment]
Nice 👌
Feel free to also push a generated mock if you want

**Comment by joschahenningsen** [comment]
Sorry for taking ages to come to this :see_no_evil: 
Looks great, I'll write a hint for profs and link to the docs. Only thing I changed is that the sections scroll horizontally like this because the vertical scrolling was a little broken depending on if there was a description or not: 
image

**Review by joschahenningsen — APPROVED** [review]

**Pull request merged**